### DISCUSSION TITLE
Sharing UI Compose code between Android and desktop

### DISCUSSION THREAD

**Discussion by opatry** [joreilly/PeopleInSpace#42] [General]
Hi 👋 .

First, many thanks for open sourcing this project, really helping new comers working on Kotlin Multiplatform 👏 .

I noticed you went for separate UI Compose implementation for Android and desktop.
Did you considered sharing some (or most?) composables between the two?
If so, do you have pros & cons to share?

I'm thinking about something like a `:common-compose` module storing shared composables, `:app` and `:desktop-app` modules having only Android integration and Windowed application integration depending no `:common-compose`.

I'm asking this because I'm currently working on this kind of approach and I was checking `PoepleInSpace` for inspiration 😅 .

My main challenges are related to `drawable`, `strings` and `font` abstraction.
Given that I have more UI than real logic (small casual game), I'd like to share compose code as much as possible :)

So far, I tried to mimic Android API on desktop as much as possible (with the help of a Gradle task generating a `R.kt` from assets to mimic Android `R.xxx.myresource_id` use) to share as much code as possible.

**Comment by Shabinder** [comment]
@opatry I am also following the same approach as yours , Using a seperate Compose Module for using in both android and desktop.I am.also using fonts, and assets using expect actual , if you want you can check out the project here: https://github.com/Shabinder/SpotiFlyer/tree/c9696fa4aa7f99f614bc1d65ae0a8dd5f223a4af

**Reply by opatry** [reply]
I'm not sure using expect/actual for assets will be ideal for future. If everyone defines its own abstraction layer, it will be a nightmare to combine libraries, switch from one project to another etc.
Ideally, I'd like Jetbrains & Google define such abstraction.
In the meantime, I try to hijack things to get closed to Android (which I consider more main stream than CfD).

That being said, I'll check out what you achieved, might be helpful 👍

**Reply by Shabinder** [reply]
I agree , expect actual is not at all ideal , 
When I started out , I also looked at available options and asked in the community , and I was told that compose team will ve bringing a more robust asset management in maybe coming 1.0 stable release.

So till then , expect / actual is just a workaround to getting things done.

**Comment by joreilly** [comment]
Hi,

This is definitely something I'm conscious about (in this and also some other projects I have where there's even more Compose code that could be shared). I haven't really had chance yet to look at it but what you described sounds like it should work.  I think I remember seeing that IntelliJ new project wizard allows creating multiplatform compose project so might be some inspiration there as well.
